ukvm/ukvm_gdb_kvm_x86_64.c

```diff
@@ -63,60 +65,83 @@ SLIST_HEAD(breakpoints_head, breakpoint_t);
 static struct breakpoints_head sw_breakpoints;
 static struct breakpoints_head hw_breakpoints;
 
-/* Not really documented. But, QEMU uses a hardcoded max of 4. */
+/* The Intel SDM specifies that the DR7 has space for 4 breakpoints. */
 #define MAX_HW_BREAKPOINTS             4
 static uint32_t nr_hw_breakpoints = 0;
 
 /* Stepping is disabled by default. */
-static uint32_t stepping = 0;
+static bool stepping = false;
 /* This is the trap instruction used for software breakpoints. */
 static const uint8_t int3 = 0xcc;
 
 static int kvm_arch_insert_sw_breakpoint(struct ukvm_hv *hv, struct breakpoint_t *bp)
 {
-    bp->saved_insn = *((uint8_t *)hv->mem + bp->addr);
-    /* It was already checked that addr is good. */
-    *((uint8_t *)hv->mem + bp->addr) = int3;
+    /* The address check at the GDB server just returned an error if addr was
+     * bad. UKVM_CHECKED_GPA_P will panic if that's the case. */
+    uint8_t *insn = UKVM_CHECKED_GPA_P(hv, bp->addr, bp->len);
+    bp->saved_insn = *insn;
+    /*
+     * We just modify the first byte even if the instruction is multi-byte.
+     * The debugger keeps track of the length of the instruction. The
+     * consequence of this is that we don't have to set all other bytes as
+     * NOP's.
+     */
+    *insn = int3;
     return 0;
 }
 
 static int kvm_arch_remove_sw_breakpoint(struct ukvm_hv *hv, struct breakpoint_t *bp)
 {
-    assert((uint8_t) *((uint8_t *)hv->mem + bp->addr) == int3);
-    /* It was already checked that addr is good. */
-    *((uint8_t *)hv->mem + bp->addr) = bp->saved_insn;
+    uint8_t *insn = UKVM_CHECKED_GPA_P(hv, bp->addr, bp->len);
+    assert(*insn == int3);
+    *insn = bp->saved_insn;
     return 0;
 }
 
 static int ukvm_gdb_update_guest_debug(struct ukvm_hv *hv)
 {
     struct kvm_guest_debug dbg = {0};
     struct breakpoint_t *bp;
-
     const uint8_t type_code[] = {
+        /* Break on instruction execution only. */
         [GDB_BREAKPOINT_HW] = 0x0,
+        /* Break on data writes only. */
         [GDB_WATCHPOINT_WRITE] = 0x1,
+        /* Break on data reads or writes but not instruction fetches. */
         [GDB_WATCHPOINT_ACCESS] = 0x3
     };
     const uint8_t len_code[] = {
+        /*
+         * 00 — 1-byte length.
+         * 01 — 2-byte length.
+         * 10 — 8-byte length.
+         * 11 — 4-byte length.
+         */
         [1] = 0x0, [2] = 0x1, [4] = 0x3, [8] = 0x2
     };
     int n = 0;
 
     if (stepping)
         dbg.control = KVM_GUESTDBG_ENABLE | KVM_GUESTDBG_SINGLESTEP;
 
-    if (!SLIST_EMPTY(&sw_breakpoints)) {
+    if (!SLIST_EMPTY(&sw_breakpoints))
         dbg.control |= KVM_GUESTDBG_ENABLE | KVM_GUESTDBG_USE_SW_BP;
-    } else {
```

Comment:
This was a bug, we can have sw and hw at the same time. Fixed in the last commit.